core/routing_prophet.go

```diff
@@ -0,0 +1,532 @@
+				"peer":        peerID,
+				"ownPred":     ownPred,
+				"peerPred":    peerPred,
+			}).Debug("Peer is not good forwarding candidate")
+		}
+	}
+
+	if len(sender) == 0 {
+		log.WithFields(
+			log.Fields{
+				"bundle": bndl.ID(),
+			}).Debug("Did not find peer to forward to")
+		return
+	}
+
+	bundleItem.Properties["routing/prophet/sent"] = sentEids
+	if err := prophet.c.store.Update(bundleItem); err != nil {
+		log.WithFields(log.Fields{
+			"error": err.Error(),
+		}).Warn("Updating BundleItem failed")
+	}
+
+	log.WithFields(log.Fields{
+		"bundle":              bndl.ID(),
+		"sent":                sentEids,
+		"convergence-senders": sender,
+	}).Debug("Prophet selected Convergence Senders for an outgoing bundle")
+
+	return
+}
+
+func (prophet *Prophet) ReportFailure(bp BundlePack, sender cla.ConvergenceSender) {
+	// When a transmission fails, that's unfortunate bet there really is not a whole lot to do
+}
+
+func (prophet *Prophet) ReportPeerAppeared(peer cla.Convergence) {
+	log.WithFields(log.Fields{
+		"address": peer,
+	}).Debug("Peer appeared")
+
+	peerReceiver, ok := peer.(cla.ConvergenceSender)
+	if !ok {
+		log.Warn("Peer was not a ConvergenceSender")
+		return
+	}
+
+	peerID := peerReceiver.GetPeerEndpointID()
+
+	log.WithFields(log.Fields{
+		"peer": peerID,
+	}).Debug("PeerID discovered")
+
+	// update our delivery predictability for this peer
+	prophet.dataMutex.Lock()
+	prophet.encounter(peerID)
+	prophet.dataMutex.Unlock()
+
+	// send them our summary vector
+	prophet.sendMetadata(peerID)
+}
+
+func (prophet *Prophet) ReportPeerDisappeared(peer cla.Convergence) {
+	log.WithFields(log.Fields{
+		"address": peer,
+	}).Debug("Peer disappeared")
+	// again, there really isn't anything to do upon a peer's disappearance
+}
+
+// TODO: Turn this into an administrative record
+
+const ExtBlockTypeProphetBlock uint64 = 194
+
+// DTLSRBlock contains routing metadata
+type ProphetBlock predictabilities
+
+func newProphetBlock(data predictabilities) *ProphetBlock {
+	newBlock := ProphetBlock(data)
+	return &newBlock
+}
+
+func (pBlock *ProphetBlock) getPredictabilities() predictabilities {
+	return predictabilities(*pBlock)
+}
+
+func (pBlock *ProphetBlock) BlockTypeCode() uint64 {
+	return ExtBlockTypeProphetBlock
+}
+
+func (pBlock ProphetBlock) CheckValid() error {
+	return nil
+}
+
+func (pBlock *ProphetBlock) MarshalCbor(w io.Writer) error {
+	// start with the outer array
+	if err := cboring.WriteArrayLength(2, w); err != nil {
+		return err
+	}
+
+	// write endpoint id
+	if err := cboring.Marshal(&pBlock.id, w); err != nil {
+		return err
+	}
+
+	// write the peer data array header
+	if err := cboring.WriteArrayLength(uint64(len(pBlock.predictability)), w); err != nil {
+		return err
+	}
+
+	// write the actual data
+	for peerID, pred := range pBlock.predictability {
+		if err := cboring.Marshal(&peerID, w); err != nil {
+			return err
+		}
+		if err := cboring.WriteFloat64(pred, w); err != nil {
+			return err
+		}
+	}
```

Comment:
First you create an array of length `len(pBlock.predictability)`, but then you write `2 * len(pBlock.predictability)` entries. So either change this to a CBOR map or double the length.

Reply:
We had a discussion about marshalling maps when I was implementing dtlsr, and I thought we had come to the conclusion that the easiest way would be to consider each key-value pair as one item and write them in one array

Reply:
This is how CBOR maps are working. So you could either change the initial CBOR tag from the array type to the map type or just double the length. However, the current state produces invalid CBOR and I'd like to avoid this.

Reply:
I see, I wasn't aware that you had implemented a designated Map-Type in cboring.

So, if I change the `cboring.WriteArrayLength` call to `cboring.WriteMapPairLength`, will that do the trick, or do I also need to further change the way I marshal the individual entries?

Reply:
Changing the CBOR type by exchanging the serialization/deserialization function should do the trick. For reference, check cboring's [map tests](https://github.com/dtn7/cboring/blob/v0.1.4/primitives_test.go#L173) and compare them against the [array tests](https://github.com/dtn7/cboring/blob/v0.1.4/primitives_test.go#L140).

Reply:
Okay, I also made the same change to the DTLSR implementation